stringfuzz/generators/clever.py

```diff
@@ -0,0 +1,356 @@
+
+
+class LibInserter(ASTWalker):
+    def __init__(self, ast, sort):
+        super().__init__(ast)
+        self.sort           = sort
+        self.candidates     = {}
+
+    def walk_expression(self, expression, parent, depth=0, pos_in_parent=-1):
+        if expression.returns(self.sort) and pos_in_parent > -1:
+            if not depth in self.candidates:
+                self.candidates[depth] = []
+            self.candidates[depth].append((parent, pos_in_parent))
+
+        for i in sorted(range(len(expression.body)), key=lambda k: random.random()):
+            sub_expression = expression.body[i]
+            if isinstance(sub_expression, ExpressionNode):
+                self.walk_expression(sub_expression, expression, depth+1, i)
+    
+    def insert_lib_calls(self, name, sig, variables, num_lib_calls):
+        min_depth = -1
+        for _ in range(num_lib_calls):
+            depth = random.choice(list(self.candidates.keys()))
+            c = random.choice(list(range(len(self.candidates[depth]))))
+            parent, pos = self.candidates[depth][c]
+            del self.candidates[depth][c]
+
+            if self.candidates[depth] == []:
+                del self.candidates[depth]
+
+            if depth < min_depth or min_depth == -1:
+                min_depth = depth
+
+            parent.body[pos] = call_func(name, sig, variables)
+        
+        return min_depth
+
+
+class LibRenamer(ASTWalker):
+    def __init__(self, ast, old_name, new_name):
+        super().__init__(ast)
+        self.old_name = old_name
+        self.new_name = new_name
+
+    def enter_identifier(self, identifier, parent):
+        if identifier.name == self.old_name:
+            identifier.name = self.new_name
+
+    def enter_expression(self, expression, parent):
+        if expression._symbol.name == self.old_name:
+            expression._symbol.name = self.new_name
+
+class Slicer(ASTWalker):
+    def __init__(self, ast, name):
+        super().__init__(ast)
+        self.name      = name
+        self.conds     = []
+        self.new_body  = None
+
+    def walk_expression(self, expression, parent):
+        if isinstance(expression, IfThenElseNode):
+            if self.name in str(expression.body[0]):
+                self.new_body = expression
+            elif self.name in str(expression.body[1]):
+                self.conds.append(expression.body[0])
+                self.walk_expression(expression.body[1], expression)
+            elif self.name in str(expression.body[2]):
+                self.conds.append(NotNode(expression.body[0]))
+                self.walk_expression(expression.body[2], expression)
+        else:
+            self.new_body = expression
+
+def call_func(name, signature, variables):
+    if isinstance(variables, dict):
+        return GenericExpressionNode(name, *[random.choice(variables[s]) for s in signature])
+    return GenericExpressionNode(name, *variables)
+
+def get_all_returning_a(sort, nodes):
```

Comment:
Are the next ~60 lines copied from `random_ast`? 😉

If so, please either reuse them, or put them into `util.py`.